File: blog-backend/app/routers/memos.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import List, Optional

from app import models
from app.models.user import UserRole
from app.core import security
from app.core.database import get_db
from app.utils.logging import get_logger

# 初始化日志记录器
logger = get_logger(__name__)
from app.schemas.memo import (
    MemoCreate, MemoUpdate, MemoResponse, MemoWithoutContent, MemoWithUser,
    MemoPasswordVerify, MemoPasswordVerifyResponse
)
from app.services.memo_service import MemoService
from app.services.activity_service import ActivityService

router = APIRouter(prefix="/memos", tags=["memos"])

@router.post("/", response_model=MemoResponse)
async def create_memo(
    memo_data: MemoCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    创建新备忘录。

    参数:
    - **title**: 备忘录标题
    - **content**: 备忘录内容
    - **is_encrypted**: 是否加密
    - **password**: 如果加密，提供密码

    返回创建的备忘录信息。
    """
    # 创建备忘录
    memo = MemoService.create_memo(db, memo_data, current_user.id)

    # 记录活动
    ActivityService.create_activity(
        db=db,
        action_type="memo_create",
        user_id=current_user.id,
        target_id=memo.id,
        description=f"用户 {current_user.username} 创建了备忘录 《{memo.title}》"
    )

    return memo

@router.get("/", response_model=List[MemoWithoutContent])
async def get_memos(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    获取当前用户的备忘录列表。

    参数:
    - **skip**: 跳过的记录数
    - **limit**: 返回的记录数

    返回备忘录列表，包含内容（除非是加密的）。
    """
    memos = MemoService.get_memos_by_user(db, current_user.id, skip, limit)

    # 处理加密内容
    result_memos = []
    for memo in memos:
        # 如果备忘录加密，清空内容
        if memo.is_encrypted:
            # 创建一个新的备忘录对象，避免修改数据库中的对象
            memo_dict = {
                "id": memo.id,
                "title": memo.title,
                "content": "",  # 清空内容
                "is_encrypted": memo.is_encrypted,
                "user_id": memo.user_id,
                "created_at": memo.created_at,
                "updated_at": memo.updated_at
            }
            result_memos.append(memo_dict)
        else:
            result_memos.append(memo)

    return result_memos

@router.get("/public", response_model=List[MemoWithUser])
async def get_public_memos(
    skip: int = Query(0, description="跳过的记录数"),
    limit: int = Query(100, description="返回的记录数"),
    db: Session = Depends(get_db)
):
    """
    获取所有公开的备忘录列表，无需登录。

    参数:
    - **skip**: 跳过的记录数
    - **limit**: 返回的记录数

    返回公开的备忘录列表，包含内容和用户信息（所有公开备忘录都是未加密的）。
    """
    # 获取所有公开的备忘录（未加密的）
    memos = MemoService.get_public_memos(db, skip, limit)

    # 添加用户信息
    result_memos = []
    for memo in memos:
        # 创建一个新的备忘录对象，包含用户信息
        memo_dict = {
            "id": memo.id,
            "title": memo.title,
            "content": memo.content,
            "is_encrypted": memo.is_encrypted,
            "user_id": memo.user_id,
            "created_at": memo.created_at,
            "updated_at": memo.updated_at,
            "user": memo.user  # 添加用户信息
        }
        result_memos.append(memo_dict)

    return result_memos

@router.get("/all", response_model=List[MemoWithUser])
async def get_all_memos(
    skip: int = Query(0, description="跳过的记录数"),
    limit: int = Query(100, description="返回的记录数"),
    db: Session = Depends(get_db)
):
    """
    获取所有备忘录列表（包括加密的），无需登录。

    参数:
    - **skip**: 跳过的记录数
    - **limit**: 返回的记录数

    返回所有备忘录列表，包括加密的。对于加密备忘录，不返回内容。
    """
    # 获取所有备忘录（包括加密的）
    memos = MemoService.get_all_memos(db, skip, limit)

    # 处理加密内容
    result_memos = []
    for memo in memos:
        # 如果备忘录加密，清空内容
        if memo.is_encrypted:
            # 创建一个新的备忘录对象，避免修改数据库中的对象
            # 只加密内容，不加密标题
            memo_dict = {
                "id": memo.id,
                "title": memo.title,  # 保留标题
                "content": "",  # 清空内容
                "is_encrypted": memo.is_encrypted,
                "user_id": memo.user_id,
                "created_at": memo.created_at,
                "updated_at": memo.updated_at,
                "user": memo.user  # 添加用户信息
            }
            result_memos.append(memo_dict)
        else:
            # 创建一个新的备忘录对象，包含用户信息
            memo_dict = {
                "id": memo.id,
                "title": memo.title,
                "content": memo.content,
                "is_encrypted": memo.is_encrypted,
                "user_id": memo.user_id,
                "created_at": memo.created_at,
                "updated_at": memo.updated_at,
                "user": memo.user  # 添加用户信息
            }
            result_memos.append(memo_dict)

    return result_memos

@router.get("/search", response_model=List[MemoWithoutContent])
async def search_memos(
    q: str = Query(..., min_length=1, description="搜索关键词"),
    skip: int = Query(0, description="跳过的记录数"),
    limit: int = Query(100, description="返回的记录数"),
    db: Session = Depends(get_db),
    current_user: Optional[models.User] = Depends(security.get_current_user_optional)
):
    """
    搜索备忘录。

    参数:
    - **q**: 搜索关键词
    - **skip**: 跳过的记录数
    - **limit**: 返回的记录数

    返回符合搜索条件的备忘录列表。
    - 如果用户已登录，则搜索该用户的所有备忘录（包括加密的）和所有公开的备忘录
    - 如果用户未登录，则只搜索公开的备忘录
    """
    # 记录搜索请求
    user_id = current_user.id if current_user else None
    logger.info(f"用户 {user_id or '未登录'} 搜索备忘录，关键词: {q}")

    # 如果用户已登录，搜索该用户的所有备忘录和所有公开的备忘录
    if current_user:
        # 搜索用户自己的备忘录（包括加密的）
        user_memos = MemoService.search_memos(
            db=db,
            query=q,
            user_id=current_user.id,
            skip=0,
            limit=limit,
            include_encrypted=True
        )

        # 搜索其他用户的公开备忘录
        public_memos = MemoService.search_memos(
            db=db,
            query=q,
            user_id=None,
            skip=0,
            limit=limit,
            include_encrypted=False
        )

        # 合并结果并去重
        memo_ids = set()
        result_memos = []

        # 先添加用户自己的备忘录
        for memo in user_memos:
            if memo.id not in memo_ids:
                memo_ids.add(memo.id)
                # 如果是加密备忘录，清空内容
                if memo.is_encrypted:
                    memo_dict = {
                        "id": memo.id,
                        "title": memo.title,  # 保留标题
                        "content": "",  # 清空内容
                        "is_encrypted": memo.is_encrypted,
                        "user_id": memo.user_id,
                        "created_at": memo.created_at,
                        "updated_at": memo.updated_at,
                        "user": memo.user  # 添加用户信息
                    }
                    result_memos.append(memo_dict)
                else:
                    result_memos.append(memo)

        # 再添加其他用户的公开备忘录
        for memo in public_memos:
            if memo.id not in memo_ids and memo.user_id != current_user.id:
                memo_ids.add(memo.id)
                result_memos.append(memo)

        # 按创建时间排序
        result_memos.sort(key=lambda x: x["created_at"] if isinstance(x, dict) else x.created_at, reverse=True)

        # 应用分页
        start = skip
        end = skip + limit
        return result_memos[start:end]
    else:
        # 如果用户未登录，只搜索公开的备忘录
        memos = MemoService.search_memos(
            db=db,
            query=q,
            user_id=None,
            skip=skip,
            limit=limit,
            include_encrypted=False
        )
        return memos

@router.get("/public/{memo_id}", response_model=MemoResponse)
async def get_public_memo(
    memo_id: int,
    db: Session = Depends(get_db)
):
    """
    获取指定ID的公开备忘录，无需登录。

    参数:
    - **memo_id**: 备忘录ID

    返回备忘录信息。如果备忘录加密，则返回404错误。
    """
    # 获取备忘录
    memo = MemoService.get_memo_by_id(db, memo_id)
    if not memo:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="备忘录不存在"
        )

    # 检查是否是公开备忘录
    if memo.is_encrypted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="备忘录不存在或已加密"
        )

    return memo

@router.get("/{memo_id}", response_model=MemoResponse)
@router.get("/{memo_id}/", response_model=MemoResponse)  # 添加带斜杠的路由
async def get_memo(
    memo_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    获取指定ID的备忘录。

    参数:
    - **memo_id**: 备忘录ID

    返回备忘录信息。如果备忘录加密，则不返回内容。
    """
    # 获取备忘录
    memo = MemoService.get_memo_by_id(db, memo_id)
    if not memo:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="备忘录不存在"
        )

    # 检查权限
    if memo.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="无权查看此备忘录"
        )

    # 如果备忘录加密，清空内容
    if memo.is_encrypted:
        # 创建一个新的备忘录对象，避免修改数据库中的对象
        # 只加密内容，不加密标题
        response_memo = MemoResponse(
            id=memo.id,
            title=memo.title,  # 保留标题
            content="",  # 清空内容
            is_encrypted=memo.is_encrypted,
            user_id=memo.user_id,
            created_at=memo.created_at,
            updated_at=memo.updated_at
        )
        return response_memo

    return memo

@router.post("/{memo_id}/verify", response_model=MemoPasswordVerifyResponse)
@router.post("/{memo_id}/verify/", response_model=MemoPasswordVerifyResponse)  # 添加带斜杠的路由
async def verify_memo_password(
    memo_id: int,
    password_data: MemoPasswordVerify,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    验证备忘录密码。

    参数:
    - **memo_id**: 备忘录ID
    - **password**: 密码

    返回验证结果和备忘录内容（如果验证成功）。
    任何已登录用户都可以尝试验证密码，只要密码正确就能查看内容。
    """
    # 获取备忘录
    memo = MemoService.get_memo_by_id(db, memo_id)
    if not memo:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="备忘录不存在"
        )

    # 不再检查用户是否是备忘录的创建者
    # 任何已登录用户都可以尝试验证密码
    # 密码验证成功与否由 verify_memo_password 函数决定

    # 记录尝试验证密码的用户信息
    logger.info(f"用户 {current_user.id} ({current_user.username}) 尝试验证备忘录 {memo_id} 的密码")

    # 验证密码
    verified_memo = MemoService.verify_memo_password(db, memo_id, password_data.password)
    if not verified_memo:
        logger.warning(f"用户 {current_user.id} 验证备忘录 {memo_id} 的密码失败")
        return MemoPasswordVerifyResponse(success=False)

    # 验证成功，返回内容
    logger.info(f"用户 {current_user.id} 成功验证备忘录 {memo_id} 的密码")
    return MemoPasswordVerifyResponse(
        success=True,
        content=verified_memo.content
    )

@router.put("/{memo_id}", response_model=MemoResponse)
@router.put("/{memo_id}/", response_model=MemoResponse)  # 添加带斜杠的路由
async def update_memo(
    memo_id: int,
    memo_data: MemoUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    更新备忘录。

    参数:
    - **memo_id**: 备忘录ID
    - **title**: 备忘录标题
    - **content**: 备忘录内容
    - **is_encrypted**: 是否加密
    - **password**: 如果加密，提供密码

    返回更新后的备忘录信息。
    """
    # 检查用户是否为管理员
    is_admin = current_user.role == UserRole.admin
    logger.info(f"用户 {current_user.id} ({current_user.username}) 尝试更新备忘录 {memo_id}, 角色: {current_user.role}, 是否管理员: {is_admin}")

    # 更新备忘录
    memo = MemoService.update_memo(db, memo_id, memo_data, current_user.id, is_admin)

    # 记录活动
    ActivityService.create_activity(
        db=db,
        action_type="memo_update",
        user_id=current_user.id,
        target_id=memo.id,
        description=f"用户 {current_user.username} 更新了备忘录 《{memo.title}》"
    )

    return memo

@router.delete("/{memo_id}", status_code=status.HTTP_204_NO_CONTENT)
@router.delete("/{memo_id}/", status_code=status.HTTP_204_NO_CONTENT)  # 添加带斜杠的路由
async def delete_memo(
    memo_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    删除备忘录。

    参数:
    - **memo_id**: 备忘录ID

    返回204 No Content。
    """
    # 添加调试日志
    logger.info(f"尝试删除备忘录 ID: {memo_id}, 当前用户 ID: {current_user.id}, 用户名: {current_user.username}, 角色: {current_user.role}")

    # 获取备忘录信息用于活动记录
    memo = MemoService.get_memo_by_id(db, memo_id)
    if not memo:
        logger.warning(f"备忘录不存在: {memo_id}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="备忘录不存在"
        )

    # 添加备忘录信息日志
    logger.info(f"找到备忘录: ID={memo.id}, 标题={memo.title}, 创建者ID={memo.user_id}")

    # 检查权限 - 允许管理员或创建者删除
    is_admin = current_user.role == "admin"
    is_owner = memo.user_id == current_user.id

    if not (is_admin or is_owner):
        logger.warning(f"权限拒绝: 用户 {current_user.id} 尝试删除备忘录 {memo_id}, 但该备忘录属于用户 {memo.user_id}")
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="无权删除此备忘录"
        )

    # 记录活动
    ActivityService.create_activity(
        db=db,
        action_type="memo_delete",
        user_id=current_user.id,
        target_id=memo_id,
        description=f"用户 {current_user.username} 删除了备忘录 《{memo.title}》"
    )

    # 调用服务层方法删除备忘录，传递管理员状态
    is_admin = current_user.role == "admin"
    MemoService.delete_memo(db, memo_id, current_user.id, is_admin=is_admin)

File: blog-frontend/src/api/modules/memos.js
/**
 * 备忘录相关 API
 * 封装所有与备忘录数据相关的网络请求
 */

import apiClient from '../core/apiClient';
import { API_PATHS } from '../core/apiPaths';

/**
 * 获取备忘录列表
 * @param {Object} params - 查询参数
 * @param {number} [params.skip=0] - 跳过的记录数
 * @param {number} [params.limit=100] - 返回的记录数
 * @returns {Promise} 返回包含备忘录列表的Promise对象
 */
export const getMemos = async (params = {}) => {
  try {
    console.log('开始获取备忘录列表');

    // 设置查询参数
    const queryParams = {
      skip: params.skip || 0,
      limit: params.limit || 100
    };

    // 发送请求到备忘录列表端点
    const response = await apiClient.get(API_PATHS.MEMOS.BASE, {
      params: queryParams
    });

    console.log('获取备忘录列表成功:', response);
    return response;
  } catch (error) {
    console.error('获取备忘录列表失败:', error);
    throw error;
  }
};

/**
 * 获取指定ID的备忘录
 * @param {number} memoId - 备忘录ID
 * @returns {Promise} 返回包含备忘录信息的Promise对象
 */
export const getMemo = async (memoId) => {
  try {
    console.log(`开始获取备忘录 ${memoId}`);

    // 发送请求到备忘录详情端点
    const response = await apiClient.get(API_PATHS.MEMOS.BY_ID(memoId));

    console.log(`获取备忘录 ${memoId} 成功:`, response);
    return response;
  } catch (error) {
    console.error(`获取备忘录 ${memoId} 失败:`, error);
    throw error;
  }
};

/**
 * 创建新备忘录
 * @param {Object} memoData - 备忘录数据
 * @param {string} memoData.title - 备忘录标题
 * @param {string} memoData.content - 备忘录内容
 * @param {boolean} [memoData.is_encrypted=false] - 是否加密
 * @param {string} [memoData.password] - 如果加密，提供密码
 * @returns {Promise} 返回包含创建的备忘录信息的Promise对象
 */
export const createMemo = async (memoData) => {
  try {
    console.log('开始创建备忘录，发送的数据:', JSON.stringify(memoData, null, 2));

    // 发送请求到创建备忘录端点
    const response = await apiClient.post(API_PATHS.MEMOS.BASE, memoData);

    console.log('创建备忘录成功，服务器响应:', response);
    return response;
  } catch (error) {
    console.error('创建备忘录失败:', error);
    console.error('错误详情:', error.response ? error.response.data : '无详细信息');
    throw error;
  }
};

/**
 * 更新备忘录
 * @param {number} memoId - 备忘录ID
 * @param {Object} memoData - 备忘录数据
 * @param {string} [memoData.title] - 备忘录标题
 * @param {string} [memoData.content] - 备忘录内容
 * @param {boolean} [memoData.is_encrypted] - 是否加密
 * @param {string} [memoData.password] - 如果加密，提供密码
 * @returns {Promise} 返回包含更新后的备忘录信息的Promise对象
 */
export const updateMemo = async (memoId, memoData) => {
  try {
    console.log(`开始更新备忘录 ${memoId}`);

    // 发送请求到更新备忘录端点
    const response = await apiClient.put(API_PATHS.MEMOS.BY_ID(memoId), memoData);

    console.log(`更新备忘录 ${memoId} 成功:`, response);
    return response;
  } catch (error) {
    console.error(`更新备忘录 ${memoId} 失败:`, error);
    throw error;
  }
};

/**
 * 删除备忘录
 * @param {number} memoId - 备忘录ID
 * @returns {Promise} 返回删除结果
 */
export const deleteMemo = async (memoId) => {
  try {
    console.log(`开始删除备忘录 ${memoId}`);

    // 发送请求到删除备忘录端点
    // 使用正确的API路径
    const url = API_PATHS.MEMOS.BY_ID(memoId);
    console.log(`删除备忘录的URL: ${url}`);
    const response = await apiClient.delete(url);

    console.log(`删除备忘录 ${memoId} 成功:`, response);
    return response;
  } catch (error) {
    console.error(`删除备忘录 ${memoId} 失败:`, error);
    throw error;
  }
};

/**
 * 验证备忘录密码
 * @param {number} memoId - 备忘录ID
 * @param {string} password - 密码
 * @returns {Promise} 返回验证结果和备忘录内容（如果验证成功）
 */
export const verifyMemoPassword = async (memoId, password) => {
  try {
    console.log(`开始验证备忘录 ${memoId} 的密码`);

    // 发送请求到验证密码端点
    const response = await apiClient.post(API_PATHS.MEMOS.VERIFY(memoId), {
      password
    });

    console.log(`验证备忘录 ${memoId} 的密码成功:`, response);
    return response;
  } catch (error) {
    console.error(`验证备忘录 ${memoId} 的密码失败:`, error);
    throw error;
  }
};

/**
 * 获取公开备忘录列表，无需登录
 * @param {Object} params - 查询参数
 * @param {number} [params.skip=0] - 跳过的记录数
 * @param {number} [params.limit=100] - 返回的记录数
 * @returns {Promise} 返回包含公开备忘录列表的Promise对象
 */
export const getPublicMemos = async (params = {}) => {
  try {
    console.log('开始获取公开备忘录列表');

    // 设置查询参数
    const queryParams = {
      skip: params.skip || 0,
      limit: params.limit || 100
    };

    // 发送请求到公开备忘录列表端点
    const response = await apiClient.get(`${API_PATHS.MEMOS.BASE}public`, {
      params: queryParams
    });

    console.log('获取公开备忘录列表成功:', response);
    return response;
  } catch (error) {
    console.error('获取公开备忘录列表失败:', error);
    throw error;
  }
};

/**
 * 获取所有备忘录列表（包括加密的），无需登录
 * @param {Object} params - 查询参数
 * @param {number} [params.skip=0] - 跳过的记录数
 * @param {number} [params.limit=100] - 返回的记录数
 * @returns {Promise} 返回包含所有备忘录列表的Promise对象
 */
export const getAllMemos = async (params = {}) => {
  try {
    console.log('开始获取所有备忘录列表（包括加密的）');

    // 设置查询参数
    const queryParams = {
      skip: params.skip || 0,
      limit: params.limit || 100
    };

    // 发送请求到所有备忘录列表端点
    const response = await apiClient.get(`${API_PATHS.MEMOS.BASE}all`, {
      params: queryParams
    });

    console.log('获取所有备忘录列表成功:', response);
    return response;
  } catch (error) {
    console.error('获取所有备忘录列表失败:', error);
    throw error;
  }
};



/**
 * 获取指定ID的公开备忘录，无需登录
 * @param {number} memoId - 备忘录ID
 * @returns {Promise} 返回包含公开备忘录信息的Promise对象
 */
export const getPublicMemo = async (memoId) => {
  try {
    console.log(`开始获取公开备忘录 ${memoId}`);

    // 发送请求到公开备忘录详情端点
    const response = await apiClient.get(`${API_PATHS.MEMOS.BASE}public/${memoId}`);

    console.log(`获取公开备忘录 ${memoId} 成功:`, response);
    return response;
  } catch (error) {
    console.error(`获取公开备忘录 ${memoId} 失败:`, error);
    throw error;
  }
};

/**
 * 搜索备忘录
 * @param {Object} params - 搜索参数
 * @param {string} params.q - 搜索关键词
 * @param {number} [params.skip=0] - 跳过的记录数
 * @param {number} [params.limit=100] - 返回的记录数
 * @returns {Promise} 返回包含搜索结果的Promise对象
 */
export const searchMemos = async (params) => {
  try {
    if (!params.q) {
      throw new Error('搜索关键词不能为空');
    }

    console.log('开始搜索备忘录，关键词:', params.q);

    // 设置查询参数
    const queryParams = {
      q: params.q,
      skip: params.skip || 0,
      limit: params.limit || 100
    };

    // 发送请求到搜索端点
    const response = await apiClient.get(API_PATHS.MEMOS.SEARCH, {
      params: queryParams
    });

    console.log('搜索备忘录成功:', response);
    return response;
  } catch (error) {
    console.error('搜索备忘录失败:', error);
    throw error;
  }
};

export default {
  getMemos,
  getMemo,
  createMemo,
  updateMemo,
  deleteMemo,
  verifyMemoPassword,
  getPublicMemos,
  getAllMemos,
  getPublicMemo,
  searchMemos
};
